File: sdks/python/agentreel/recorder.py
"""Recorder — wraps LLM API calls and builds a trajectory automatically."""

from __future__ import annotations

import time
from datetime import datetime, timezone
from typing import Any

from agentreel.trajectory import (
    LlmCall,
    Message,
    Step,
    TokenUsage,
    ToolCall,
    Trajectory,
)


class Recorder:
    """Records LLM interactions into a Trajectory.

    Usage:
        recorder = Recorder(title="My agent run")

        # Record an LLM call
        recorder.record_llm_call(
            model="gpt-4o",
            messages=[{"role": "user", "content": "Hello"}],
            response={"choices": [{"message": {"role": "assistant", "content": "Hi!"}}]},
            duration_ms=500,
        )

        # Get the trajectory
        trajectory = recorder.finalize()
        trajectory.save("my_run.trajectory.json")
    """

    def __init__(
        self,
        title: str | None = None,
        tags: list[str] | None = None,
        trajectory: Trajectory | None = None,
    ):
        self.trajectory = trajectory or Trajectory()
        if title:
            self.trajectory.metadata.title = title
        if tags:
            self.trajectory.metadata.tags = tags

    def record_llm_call(
        self,
        model: str,
        messages: list[dict[str, Any]],
        response: dict[str, Any],
        duration_ms: float | None = None,
    ) -> int:
        """Record an LLM API call and return the step index."""
        # Extract response content
        resp_content = _extract_response_content(response)
        tokens = _extract_tokens(response)
        stop_reason = _extract_stop_reason(response)

        step = Step(
            type="llm_call",
            timestamp=_now_iso(),
            duration_ms=duration_ms,
            llm_call=LlmCall(
                model=model,
                messages=[Message(role=m.get("role", "user"), content=m.get("content")) for m in messages],
                response=Message(role="assistant", content=resp_content),
                stop_reason=stop_reason,
            ),
            tokens=tokens,
        )

        return self.trajectory.add_step(step)

    def record_tool_call(
        self,
        name: str,
        input: dict[str, Any] | None = None,
        output: Any = None,
        status: str = "success",
        duration_ms: float | None = None,
    ) -> int:
        """Record a tool call and return the step index."""
        step = Step(
            type="tool_call",
            timestamp=_now_iso(),
            duration_ms=duration_ms,
            tool_call=ToolCall(
                name=name,
                input=input,
                output=output,
                status=status,
            ),
        )
        return self.trajectory.add_step(step)

    def record_error(
        self,
        message: str,
        code: str | None = None,
        recoverable: bool = True,
    ) -> int:
        """Record an error and return the step index."""
        from agentreel.trajectory import Error

        step = Step(
            type="error",
            timestamp=_now_iso(),
            error=Error(code=code, message=message, recoverable=recoverable),
        )
        return self.trajectory.add_step(step)

    def finalize(self, status: str = "success", summary: str | None = None) -> Trajectory:
        """Finalize and return the trajectory with computed stats."""
        from agentreel.trajectory import Outcome

        self.trajectory.metadata.completed_at = _now_iso()
        self.trajectory.outcome = Outcome(status=status, summary=summary)
        self.trajectory.compute_stats()
        return self.trajectory


def _extract_response_content(response: dict[str, Any]) -> str | None:
    # OpenAI format
    choices = response.get("choices", [])
    if choices:
        msg = choices[0].get("message", {})
        return msg.get("content")

    # Anthropic format
    content = response.get("content", [])
    if content:
        texts = [b["text"] for b in content if b.get("type") == "text" and "text" in b]
        return "\n".join(texts) if texts else None

    return None


def _extract_tokens(response: dict[str, Any]) -> TokenUsage | None:
    usage = response.get("usage")
    if not usage:
        return None

    return TokenUsage(
        input_tokens=usage.get("input_tokens") or usage.get("prompt_tokens"),
        output_tokens=usage.get("output_tokens") or usage.get("completion_tokens"),
        cache_read_tokens=usage.get("cache_read_input_tokens"),
        cache_write_tokens=usage.get("cache_creation_input_tokens"),
    )


def _extract_stop_reason(response: dict[str, Any]) -> str | None:
    # Anthropic
    if reason := response.get("stop_reason"):
        return reason
    # OpenAI
    choices = response.get("choices", [])
    if choices:
        reason = choices[0].get("finish_reason")
        if reason == "stop":
            return "end_turn"
        if reason == "tool_calls":
            return "tool_use"
        if reason == "length":
            return "max_tokens"
        return reason
    return None


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()

File: sdks/typescript/src/recorder.ts
import { Trajectory } from "./trajectory.js";
import type { LlmCall, Message, Step, TokenUsage, ToolCall } from "./types.js";

export class Recorder {
  private trajectory: Trajectory;

  constructor(options?: { title?: string; tags?: string[]; trajectory?: Trajectory }) {
    this.trajectory = options?.trajectory ?? new Trajectory();
    if (options?.title) this.trajectory.metadata.title = options.title;
    if (options?.tags) this.trajectory.metadata.tags = options.tags;
  }

  recordLlmCall(params: {
    model: string;
    messages: Array<{ role: string; content?: string }>;
    response: Record<string, unknown>;
    duration_ms?: number;
  }): number {
    const respContent = extractResponseContent(params.response);
    const tokens = extractTokens(params.response);
    const stopReason = extractStopReason(params.response);

    return this.trajectory.addStep({
      type: "llm_call",
      duration_ms: params.duration_ms,
      llm_call: {
        model: params.model,
        messages: params.messages.map((m) => ({
          role: m.role as Message["role"],
          content: m.content,
        })),
        response: { role: "assistant", content: respContent ?? undefined },
        stop_reason: stopReason ?? undefined,
      },
      tokens: tokens ?? undefined,
    });
  }

  recordToolCall(params: {
    name: string;
    input?: Record<string, unknown>;
    output?: unknown;
    status?: "success" | "error" | "timeout" | "denied";
    duration_ms?: number;
  }): number {
    return this.trajectory.addStep({
      type: "tool_call",
      duration_ms: params.duration_ms,
      tool_call: {
        name: params.name,
        input: params.input,
        output: params.output,
        status: params.status ?? "success",
      },
    });
  }

  recordError(params: {
    message: string;
    code?: string;
    recoverable?: boolean;
  }): number {
    return this.trajectory.addStep({
      type: "error",
      error: {
        message: params.message,
        code: params.code,
        recoverable: params.recoverable ?? true,
      },
    });
  }

  finalize(options?: {
    status?: "success" | "failure" | "partial" | "aborted";
    summary?: string;
  }): Trajectory {
    this.trajectory.metadata.completed_at = new Date().toISOString();
    this.trajectory.outcome = {
      status: options?.status ?? "success",
      summary: options?.summary,
    };
    this.trajectory.computeStats();
    return this.trajectory;
  }
}

function extractResponseContent(response: Record<string, unknown>): string | null {
  // OpenAI
  const choices = response.choices as Array<Record<string, unknown>> | undefined;
  if (choices?.length) {
    const msg = choices[0].message as Record<string, unknown> | undefined;
    if (msg?.content && typeof msg.content === "string") return msg.content;
  }

  // Anthropic
  const content = response.content as Array<Record<string, unknown>> | undefined;
  if (content?.length) {
    const texts = content
      .filter((b) => b.type === "text" && typeof b.text === "string")
      .map((b) => b.text as string);
    return texts.length > 0 ? texts.join("\n") : null;
  }

  return null;
}

function extractTokens(response: Record<string, unknown>): TokenUsage | null {
  const usage = response.usage as Record<string, unknown> | undefined;
  if (!usage) return null;

  return {
    input_tokens:
      (usage.input_tokens as number) ?? (usage.prompt_tokens as number) ?? undefined,
    output_tokens:
      (usage.output_tokens as number) ?? (usage.completion_tokens as number) ?? undefined,
    cache_read_tokens: (usage.cache_read_input_tokens as number) ?? undefined,
    cache_write_tokens: (usage.cache_creation_input_tokens as number) ?? undefined,
  };
}

function extractStopReason(
  response: Record<string, unknown>
): LlmCall["stop_reason"] | null {
  // Anthropic
  if (typeof response.stop_reason === "string") {
    const map: Record<string, LlmCall["stop_reason"]> = {
      end_turn: "end_turn",
      tool_use: "tool_use",
      max_tokens: "max_tokens",
      stop_sequence: "stop_sequence",
    };
    return map[response.stop_reason] ?? null;
  }

  // OpenAI
  const choices = response.choices as Array<Record<string, unknown>> | undefined;
  if (choices?.length) {
    const reason = choices[0].finish_reason as string;
    const map: Record<string, LlmCall["stop_reason"]> = {
      stop: "end_turn",
      tool_calls: "tool_use",
      length: "max_tokens",
    };
    return map[reason] ?? null;
  }

  return null;
}
